protocol_details/rpmsg_protocol.rst

```diff
@@ -0,0 +1,89 @@
+
+.. image:: ../images/vrings_used_buffers.jpg
+
+Both “used” and “avail” ring buffers have a flags field. Its purpose is mainly to tell the writer whether he should interrupt the other core when updating the head of the ring. The same bit is used for this purpose in both “used” and “avail” ring buffers:
+
+.. image:: ../images/vrings_used_buffers_flags.jpg
+
+Transport Layer - RPMsg
+-----------------------
+RPMsg Header Definition
+~~~~~~~~~~~~~~~~~~~~~~~
+
+Each RPMsg message is contained in a buffer, which is present in the shared memory. This buffer is pointed to by the address field of a buffer descriptor from vring’s buffer descriptor pool. The first 16 bytes of this buffer are used internally by the transport layer (RPMsg layer). The first word (32bits) is used as an address of the sender or source endpoint, next word is the address of the receiver or destination endpoint. There is a reserved field for alignment reasons (RPMsg header is thus 16 bytes aligned). Last two fields of the header are the length of the payload (16bit) and a 16-bit flags field. The reserved field is not used to transmit data between cores and can be used internally in the RPMsg implementation. The user payload follows the RPMsg header.
+
+.. image:: ../images/rpmsg_header.jpg
+
+Special consideration should be taken if an alignment greater than 16 bytes is required; however, this is not typical for a shared memory, which should be fast and is therefore often not cached (alignment greater than 8 bytes is not needed at all).
+
+Flags Field
+~~~~~~~~~~~
+
+The flags field of the RPMsg header is currently unused by RPMsg and is reserved. Any propositions for what this field could be used for is welcome. It could be released for application use, but this can be considered as inconsistent - RPMsg header would not be aligned and the reserved field would be therefore useless.
+
+.. image:: ../images/rpmsg_flags.jpg
+
+RPMsg Channel
+~~~~~~~~~~~~~
+
+Every remote core in RPMsg component is represented by RPMsg device that provides a communication channel between master and remote, hence RPMsg devices are also known as channels RPMsg channel is identified by the textual name and local (source) and destination address. The RPMsg framework keeps track of channels using their names.
+
+.. _rpmsg-endpoint:
+
+RPMsg Endpoint
+~~~~~~~~~~~~~~
+
+RPMsg endpoints provide logical connections on top of RPMsg channel. It allows the user to bind multiple rx callbacks on the same channel.
+
+Every RPMsg endpoint has a unique src address and associated call back function. When an application creates an endpoint with the local address, all the further inbound messages with the destination address equal to local address of endpoint are routed to that callback function. Every channel has a default endpoint which enables applications to communicate without even creating new endpoints.
```

Comment:
s/all the further inbound/all further inbound/
s/to local address/to the local address/